src/renderers/dom/shared/__tests__/ReactDOMServerIntegration-test.js

```diff
@@ -996,4 +1021,448 @@ describe('ReactDOMServerIntegration', () => {
+
+      it('should reconnect a div with a number and string version of number', () =>
+        expectMarkupMatch(<div>{2}</div>, <div>2</div>));
+
+      it('should error reconnecting different numbers', () =>
+        expectMarkupMismatch(<div>{2}</div>, <div>{3}</div>));
+
+      it('should error reconnecting different number from text', () =>
+        expectMarkupMismatch(<div>{2}</div>, <div>3</div>));
+
+      it('should error reconnecting different text in two code blocks', () =>
+        expectMarkupMismatch(
+          <div>{'Text1'}{'Text2'}</div>,
+          <div>{'Text1'}{'Text3'}</div>,
+        ));
+    });
+
+    describe('element trees and children', function() {
+      it('should error reconnecting missing children', () =>
+        expectMarkupMismatch(<div><div /></div>, <div />));
+
+      it('should error reconnecting added children', () =>
+        expectMarkupMismatch(<div />, <div><div /></div>));
+
+      it('should error reconnecting more children', () =>
+        expectMarkupMismatch(<div><div /></div>, <div><div /><div /></div>));
+
+      it('should error reconnecting fewer children', () =>
+        expectMarkupMismatch(<div><div /><div /></div>, <div><div /></div>));
+
+      it('should error reconnecting reordered children', () =>
+        expectMarkupMismatch(
+          <div><div /><span /></div>,
+          <div><span /><div /></div>,
+        ));
+
+      it('should error reconnecting a div with children separated by whitespace on the client', () =>
+        expectMarkupMismatch(
+          <div id="parent"><div id="child1" /><div id="child2" /></div>,
+          // prettier-ignore
+          <div id="parent"><div id="child1" />      <div id="child2" /></div>, // eslint-disable-line no-multi-spaces
+        ));
+
+      it('should error reconnecting a div with children separated by different whitespace on the server', () =>
+        expectMarkupMismatch(
+          // prettier-ignore
+          <div id="parent"><div id="child1" />      <div id="child2" /></div>, // eslint-disable-line no-multi-spaces
+          <div id="parent"><div id="child1" /><div id="child2" /></div>,
+        ));
+
+      it('should error reconnecting a div with children separated by different whitespace', () =>
+        expectMarkupMismatch(
+          <div id="parent"><div id="child1" /> <div id="child2" /></div>,
+          // prettier-ignore
+          <div id="parent"><div id="child1" />      <div id="child2" /></div>, // eslint-disable-line no-multi-spaces
+        ));
+
+      it('can distinguish an empty component from a dom node', () =>
+        expectMarkupMismatch(
+          <div><span /></div>,
+          <div><EmptyComponent /></div>,
+        ));
+
+      it('can distinguish an empty component from an empty text component', () =>
+        expectMarkupMismatch(<div><EmptyComponent /></div>, <div>{''}</div>));
+
+      it('should reconnect if component trees differ but resulting markup is the same', () => {
```

Comment:
Good question! I could argue that it's a little different, in that it's comparing two different React.Component class implementations against each other, which wasn't done above, but yeah, it's probably duplicative and has a very low (maybe zero?) likelihood of catching something not caught by those tests. I can take it out in the next PR.